File: js/game.js
const { Board, Moves, MovesAsStr, ReverseMoves } = require('./board');

//FIXME disabled for now, buggy!
const PARTIAL_REPLAY_TRESHOLD = 90;

class Game {

  constructor(board) {
    this.initialBoard = board;
    this.board = board.clone();
    this.steps = [];
  }

  
  moveBoardToState(board, steps) {
    for(let step of steps) {
      board.move(step);
    }
  }

  lastSameMoveIdx(steps, lastSteps) {
    const length = Math.min(lastSteps.length, steps.length);
    for(let i = 0; i < length; i++) {
      if(lastSteps[i] !== steps[i]) {
        return i;
      }
    }
    return length - 1;
  }


  diffSteps(steps, lastSteps, lastSameMoveIdx) {
    const moves = [];
    for(let i = lastSteps.length - 1; i > lastSameMoveIdx; i--) {
      moves.push(lastSteps[i]);
    }
    for(let i = lastSameMoveIdx + 1; i < steps.length; i++) {
      moves.push(steps[i]);
    }
    return moves;
  }

  replayEntireBoard(steps) {
    this.board = this.initialBoard.clone();
    this.moveBoardToState(this.board, steps);
    this.steps = steps;
  }

  undoAndReplayPartially(steps, lastSteps, lastSameMoveIdx) {
    const diffSteps = this.diffSteps(steps, lastSteps, lastSameMoveIdx);
    this.moveBoardToState(this.board, diffSteps);
    this.steps = steps;
  }

  updateBoardForSteps(steps, lastSteps) {
    if(!lastSteps || steps.length < PARTIAL_REPLAY_TRESHOLD) {
      this.replayEntireBoard(steps);
    } else {
      const lastSameMoveIdx = this.lastSameMoveIdx(steps, lastSteps);
      const diffMoveCount = lastSteps.length - lastSameMoveIdx + steps.length - lastSameMoveIdx;
      if(diffMoveCount < steps.length) {
        this.undoAndReplayPartially(steps, lastSteps, lastSameMoveIdx);
      } else {
        this.replayEntireBoard(steps);
      }

    }
  }

  updateBoardForGrid(grid, dist, steps) {
    this.board = new Board(grid, true);
    this.steps = steps;
  }

  stepsAsStr() {
    return this.steps.map(m => MovesAsStr[m]);
  }

  //DELEGATES

  distance() {
    return this.board.dist;
  }

  move(move) {
    this.board.move(move);
  }

  undoMove(move) {
    this.board.move(ReverseMoves[move]);
  }

  solved() {
    return this.board.solved();
  }

  validMoves() {
    return this.board.validMoves();
  }

  grid() {
    return new Uint8Array(this.board.flat);
  }

}

module.exports = Game;

File: rust/fifteen-game/src/game.rs
use board::{Move, SixteenBoard};
use smallvec::SmallVec;

pub const MAX_MOVES: i32 = 80;

#[allow(dead_code)]
pub struct Game {
    initial_board: SixteenBoard,
    board: SixteenBoard,
    steps: SmallVec<[Move; 128]>,
}

impl Game {
    pub fn new(board: SixteenBoard) -> Game {
        Game {
            initial_board: board,
            board: board.clone(),
            steps: SmallVec::with_capacity(MAX_MOVES as usize)
        }
    }

    pub fn move_board_to_state(&mut self, steps: SmallVec<[Move; 128]>) {
        for step in steps {
            self.board.move_elem(step);
        }
        // TODO FIXME: this is needed
        // self.steps = steps;
    }

    pub fn replay_entire_board(&mut self, steps: SmallVec<[Move; 128]>) {
        self.board = self.initial_board.clone();
        self.move_board_to_state(steps);
    }

    pub fn distance(&self) -> i32 {
        return self.board.dist();
    }

    pub fn move_elem(&mut self, move_type: Move) {
        self.board.move_elem(move_type);
    }

    pub fn undo_move(&mut self, move_type: Move) {
        self.board.reverse_move(move_type);
    }

    pub fn solved(&self) -> bool {
        return self.board.solved();
    }

    pub fn valid_moves(&self) -> SmallVec<[Move; 4]> {
        return self.board.valid_moves();
    }
}
